File: src/pos_app/lib/features/service_requests/models/service_request.dart
import '../../../core/models/localized_text.dart';

/// Types of service requests
enum ServiceRequestType {
  callWaiter(1, 'Call Waiter'),
  controllerChange(2, 'Controller Change'),
  receiptToPay(3, 'Receipt to Pay'),
  switchToMulti(4, 'Switch to Multi'),
  switchToSingle(5, 'Switch to Single'),

  /// Switch the stay to another rate option; the option's code travels in optionCode
  changeOption(6, 'Change rate');

  final int value;
  final String label;

  const ServiceRequestType(this.value, this.label);

  static ServiceRequestType fromValue(int value) {
    return ServiceRequestType.values.firstWhere(
      (e) => e.value == value,
      orElse: () => ServiceRequestType.callWaiter,
    );
  }
}

/// Status of a service request
enum ServiceRequestStatus {
  pending(1, 'Pending'),
  acknowledged(2, 'Acknowledged'),
  completed(3, 'Completed');

  final int value;
  final String label;

  const ServiceRequestStatus(this.value, this.label);

  static ServiceRequestStatus fromValue(int value) {
    return ServiceRequestStatus.values.firstWhere(
      (e) => e.value == value,
      orElse: () => ServiceRequestStatus.pending,
    );
  }
}

/// Service request model
class ServiceRequest {
  final int id;
  final String userName;
  final int? roomId;

  /// The place: the room, or the table's name for a table request
  final LocalizedText roomName;
  final int? tableId;
  final LocalizedText? tableName;

  /// The Spaces place; 'Room', 'Table' or 'Station'
  final int? placeId;
  final String? placeKind;

  /// The rate option a changeOption request asks for
  final String? optionCode;
  final ServiceRequestType requestType;
  final ServiceRequestStatus status;
  final DateTime createdAt;

  ServiceRequest({
    required this.id,
    required this.userName,
    this.roomId,
    required this.roomName,
    this.tableId,
    this.tableName,
    this.placeId,
    this.placeKind,
    this.optionCode,
    required this.requestType,
    required this.status,
    required this.createdAt,
  });

  factory ServiceRequest.fromJson(Map<String, dynamic> json) {
    return ServiceRequest(
      id: json['id'] as int,
      userName: json['userName'] as String,
      roomId: json['roomId'] as int?,
      roomName: LocalizedText.parse(json['roomName']),
      tableId: json['tableId'] as int?,
      tableName: json['tableName'] != null ? LocalizedText.fromJson(json['tableName'] as Map<String, dynamic>) : null,
      placeId: json['placeId'] as int?,
      placeKind: json['placeKind'] as String?,
      optionCode: json['optionCode'] as String?,
      requestType: ServiceRequestType.fromValue(json['requestType'] as int),
      status: ServiceRequestStatus.fromValue(json['status'] as int),
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

File: src/Notification.API/Model/ServiceRequest.cs
namespace Ninja.Notification.API.Model;

public class ServiceRequest
{
    public int Id { get; set; }
    public string UserId { get; set; } = string.Empty;
    public string UserName { get; set; } = string.Empty;
    /// <summary>The stay the request came from; null when no clock runs at the place.</summary>
    public int? SessionId { get; set; }
    /// <summary>The Spaces place the request came from.</summary>
    public int? PlaceId { get; set; }
    /// <summary>"Room", "Table" or "Station".</summary>
    public string? PlaceKind { get; set; }
    /// <summary>The place's name as the request named it: where the till goes.</summary>
    public LocalizedText PlaceName { get; set; } = new LocalizedText(string.Empty);
    /// <summary>The rate option asked for by a <see cref="ServiceRequestType.ChangeOption"/> request ("multi", "single", ...).</summary>
    public string? OptionCode { get; set; }
    public int BranchId { get; set; }
    public ServiceRequestType RequestType { get; set; }
    public ServiceRequestStatus Status { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime? AcknowledgedAt { get; set; }
    public string? AcknowledgedBy { get; set; }
}

public enum ServiceRequestType
{
    CallWaiter = 1,
    ControllerChange = 2,
    ReceiptToPay = 3,
    // 4 and 5 were the two-option room switches, now a ChangeOption with the option's code
    /// <summary>Switch the stay to another rate option of the place's tariff; the option travels in OptionCode.</summary>
    ChangeOption = 6,
}

public enum ServiceRequestStatus
{
    Pending = 1,
    Acknowledged = 2,
    Completed = 3,
    /// <summary>Taken back by the customer before anyone picked it up.</summary>
    Cancelled = 4
}
